plugins/telink-dfu/fu-telink-dfu-hid-device.c

```diff
@@ -0,0 +1,318 @@
+		return FALSE;
+
+	/* success */
+	fu_device_sleep(FU_DEVICE(self), 20);
+	return TRUE;
+}
+
+static gboolean
+fu_telink_dfu_hid_device_ota_stop(FuTelinkDfuHidDevice *self, guint number_chunks, GError **error)
+{
+	g_autoptr(FuStructTelinkDfuEndCheck) end_check = fu_struct_telink_dfu_end_check_new();
+	g_autoptr(FuStructTelinkDfuHidPkt) pkt = NULL;
+	guint16 pkt_index = (guint16)(number_chunks >> 4) - 1;
+
+	/* last data packet index */
+	fu_struct_telink_dfu_end_check_set_pkt_index(end_check, pkt_index);
+	fu_struct_telink_dfu_end_check_set_inverted_pkt_index(end_check, ~pkt_index);
+	pkt = fu_telink_dfu_hid_device_create_packet(FU_TELINK_DFU_CMD_OTA_END,
+						     end_check->data,
+						     end_check->len,
+						     error);
+	if (!fu_telink_dfu_hid_device_write(self, pkt->data, pkt->len, error))
+		return FALSE;
+
+	/* success */
+	fu_device_sleep(FU_DEVICE(self), 10000);
+	return TRUE;
+}
+
+static gboolean
+fu_telink_dfu_hid_device_write_blob(FuTelinkDfuHidDevice *self,
+				    GBytes *blob,
+				    FuProgress *progress,
+				    GError **error)
+{
+	g_autoptr(FuChunkArray) chunks = NULL;
+	g_autoptr(FuStructTelinkDfuHidPkt) pkt = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 1, "ota-start");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 70, "ota-data");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 29, "ota-stop");
+
+	/* OTA start command */
+	if (!fu_telink_dfu_hid_device_ota_start(self, error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* OTA firmware data */
+	chunks = fu_chunk_array_new_from_bytes(blob,
+					       FU_TELINK_DFU_HID_DEVICE_START_ADDR,
+					       FU_CHUNK_PAGESZ_NONE,
+					       FU_STRUCT_TELINK_DFU_BLE_PKT_SIZE_PAYLOAD);
+	if (!fu_telink_dfu_hid_device_write_blocks(self,
+						   chunks,
+						   fu_progress_get_child(progress),
+						   error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* OTA stop command */
+	if (!fu_telink_dfu_hid_device_ota_stop(self, fu_chunk_array_length(chunks), error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* success */
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	return TRUE;
+}
+
+static gboolean
+fu_telink_dfu_hid_device_write_firmware(FuDevice *device,
+					FuFirmware *firmware,
+					FuProgress *progress,
+					FwupdInstallFlags flags,
+					GError **error)
+{
+	FuTelinkDfuHidDevice *self = FU_TELINK_DFU_HID_DEVICE(device);
+	g_autoptr(FuArchive) archive = NULL;
+	g_autoptr(GBytes) blob = NULL;
+	g_autoptr(GInputStream) stream = NULL;
+
+	/* get default image */
+	stream = fu_firmware_get_stream(firmware, error);
+	if (stream == NULL)
+		return FALSE;
+	archive = fu_archive_new_stream(stream, FU_ARCHIVE_FLAG_IGNORE_PATH, error);
+	if (archive == NULL)
+		return FALSE;
+	// FIXME, check not using .cab in fwupdtool install-blob
```

Comment:
i think you can remove the FIXME

Reply:
resolved in commit 0fb74d587e8aedf6a6bccabff38623ea034d224d